File: runtime/dart/convert/utf8Encoder.ts
import { IStream } from "../async/stream";
import { IStreamTransformer } from "../async/streamTransformer";
import { IList } from "../core/list";
import { ISink } from "../core/sink";
import { IUint8List } from "../typed_data/uint8List";
import { IConverter } from "./converter";
import { IStringConversionSink } from "./stringConversionSink";
declare const dart: {
    convert: {
        utf8Encoder: (this: void, utf8Encoder: IUtf8Encoder) => IUtf8Encoder;
    };
};
export interface IUtf8Encoder {
    convert: (
        string: string,
        start: number,
        end?: number | undefined
    ) => IUint8List;
    startChunkedConversion: (
        sink: ISink<IList<number>>
    ) => IStringConversionSink;
    bind: (stream: IStream<string>) => IStream<IList<number>>;
    fuse: <TT>(other: IConverter<IList<number>, TT>) => IConverter<string, TT>;
    cast: <RS, RT>() => IConverter<RS, RT>;
    toString: () => string;
    getHashCode: () => number;
}
export class Utf8Encoder
    implements
        IConverter<string, IList<number>>,
        IStreamTransformer<string, IList<number>>
{
    public constructor() {
        dart.convert.utf8Encoder(this);
    }
    private readonly _dart_convert: (
        string: string,
        start: number,
        end?: number | undefined
    ) => IUint8List = undefined as any;
    private readonly _dart_startChunkedConversion: (
        sink: ISink<IList<number>>
    ) => IStringConversionSink = undefined as any;
    private readonly _dart_bind: (
        stream: IStream<string>
    ) => IStream<IList<number>> = undefined as any;
    private readonly _dart_fuse: <TT>(
        other: IConverter<IList<number>, TT>
    ) => IConverter<string, TT> = undefined as any;
    private readonly _dart_cast: <RS, RT>() => IConverter<RS, RT> =
        undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    public convert(
        string: string,
        start: number = 0,
        end?: number | undefined
    ): IUint8List {
        return this._dart_convert(string, start, end);
    }
    public startChunkedConversion(
        sink: ISink<IList<number>>
    ): IStringConversionSink {
        return this._dart_startChunkedConversion(sink);
    }
    public bind(stream: IStream<string>): IStream<IList<number>> {
        return this._dart_bind(stream);
    }
    public fuse<TT>(
        other: IConverter<IList<number>, TT>
    ): IConverter<string, TT> {
        return this._dart_fuse(other);
    }
    public cast<RS, RT>(): IConverter<RS, RT> {
        return this._dart_cast();
    }
    public toString(): string {
        return this._dart_toString();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
}

File: lib/cfr/builtins/libs/dart/convert/utf8Encoder.dart
import 'dart:async';
import 'dart:convert';
import 'dart:core';
import 'dart:typed_data';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedUtf8Encoder extends VMManagedBox<Utf8Encoder> {
  VMManagedUtf8Encoder(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['convert'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Uint8List>(
            object: vmObject.convert(luaCallerArguments[1],
                luaCallerArguments[2], luaCallerArguments[3]),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['startChunkedConversion'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<StringConversionSink>(
            object: vmObject.startChunkedConversion(
                maybeUnBoxAndBuildArgument<Sink<List<int>>>(
                    luaCallerArguments[1],
                    parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['bind'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Stream>(
            object: vmObject.bind(maybeUnBoxAndBuildArgument<Stream<String>>(
                luaCallerArguments[1],
                parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['fuse'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Converter>(
            object: vmObject.fuse(
                maybeUnBoxAndBuildArgument<Converter<List<int>, dynamic>>(
                    luaCallerArguments[1],
                    parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['cast'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Converter>(
            object: vmObject.cast(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final Utf8Encoder vmObject;
}

class RTManagedUtf8Encoder extends Utf8Encoder implements Box<Utf8Encoder> {
  RTManagedUtf8Encoder({required this.table, required this.hydroState})
      : super() {
    table['vmObject'] = vmObject;
    table['unwrap'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [unwrap()];
    });
    table['_dart_convert'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Uint8List>(
            object: super.convert(luaCallerArguments[1], luaCallerArguments[2],
                luaCallerArguments[3]),
            hydroState: hydroState,
            table: HydroTable())
      ];
    });
    table['_dart_startChunkedConversion'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<StringConversionSink>(
            object: super.startChunkedConversion(
                maybeUnBoxAndBuildArgument<Sink<List<int>>>(
                    luaCallerArguments[1],
                    parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable())
      ];
    });
    table['_dart_bind'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Stream>(
            object: super.bind(maybeUnBoxAndBuildArgument<Stream<String>>(
                luaCallerArguments[1],
                parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable())
      ];
    });
    table['_dart_fuse'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Converter>(
            object: super.fuse(
                maybeUnBoxAndBuildArgument<Converter<List<int>, dynamic>>(
                    luaCallerArguments[1],
                    parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable())
      ];
    });
    table['_dart_cast'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Converter>(
            object: super.cast(), hydroState: hydroState, table: HydroTable())
      ];
    });
    table['_dart_toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.toString()];
    });
    table['_dart_getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.hashCode];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  Utf8Encoder unwrap() => this;
  Utf8Encoder get vmObject => this;
  @override
  Uint8List convert(String string, [int start = 0, int? end]) {
    Closure closure = table["convert"];
    return maybeUnBoxAndBuildArgument<Uint8List>(
        closure.dispatch([table], parentState: hydroState)[0],
        parentState: hydroState);
  }

  @override
  StringConversionSink startChunkedConversion(Sink sink) {
    Closure closure = table["startChunkedConversion"];
    return maybeUnBoxAndBuildArgument<StringConversionSink>(
        closure.dispatch([table], parentState: hydroState)[0],
        parentState: hydroState);
  }

  @override
  Stream<List<int>> bind(Stream stream) {
    Closure closure = table["bind"];
    return maybeUnBoxAndBuildArgument<Stream<List<int>>>(
        closure.dispatch([table], parentState: hydroState)[0],
        parentState: hydroState);
  }

  @override
  Converter<String, TT> fuse<TT>(Converter other) {
    Closure closure = table["fuse"];
    return maybeUnBoxAndBuildArgument<Converter<String, TT>>(
        closure.dispatch([table], parentState: hydroState)[0],
        parentState: hydroState);
  }

  @override
  Converter<RS, RT> cast<RS, RT>() {
    Closure closure = table["cast"];
    return maybeUnBoxAndBuildArgument<Converter<RS, RT>>(
        closure.dispatch([table], parentState: hydroState)[0],
        parentState: hydroState);
  }

  @override
  String toString() {
    Closure closure = table["__tostring"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  int get hashCode {
    Closure closure = table["getHashCode"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }
}

void loadUtf8Encoder(
    {required HydroState hydroState, required HydroTable table}) {
  table['utf8Encoder'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      RTManagedUtf8Encoder(table: luaCallerArguments[0], hydroState: hydroState)
    ];
  });
  registerBoxer<Utf8Encoder>(boxer: (
      {required Utf8Encoder vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedUtf8Encoder(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}
